autonomy/deploy/base.py

```diff
@@ -69,33 +68,41 @@ def __init__(
         keys: Path,
         number_of_agents: Optional[int] = None,
         private_keys_password: Optional[str] = None,
+        agent_instances: Optional[List[str]] = None,
     ) -> None:
         """Initialize the Base Deployment."""
-        self.private_keys: List = []
+        self.keys: List = []
         self.private_keys_password = private_keys_password
         self.service = load_service_config(service_path)
+
+        # we allow configurable number of agents independent of the
+        # number of agent instances for local development purposes
+
         if number_of_agents is not None:
             self.service.number_of_agents = number_of_agents
 
+        self.agent_instances = agent_instances
         self.read_keys(keys)
-        if self.service.number_of_agents > len(self.private_keys):
-            raise ValueError("Number of agents cannot be greater than available keys.")
 
     def read_keys(self, file_path: Path) -> None:
         """Read in keys from a file on disk."""
 
         keys = json.loads(file_path.read_text(encoding=DEFAULT_ENCODING))
-
-        key_schema = (
-            KEY_SCHEMA_UNENCRYPTED_KEY
-            if self.private_keys_password is None
-            else KEY_SCHEMA_ENCRYPTED_KEY
-        )
         for key in keys:
-            for required_key in [KEY_SCHEMA_ADDRESS, key_schema]:
-                if required_key not in key.keys():
-                    raise ValueError("Key file incorrectly formatted.")
-            self.private_keys.append(key[key_schema])
+            if {KEY_SCHEMA_ADDRESS, KEY_SCHEMA_PRIVATE_KEY} != set(key.keys()):
+                raise ValueError("Key file incorrectly formatted.")
+
+        if self.agent_instances is not None:
+            keys = [kp for kp in keys if kp["address"] in self.agent_instances]
+            if not keys:
+                raise ValueError(
+                    "Cannot find the provided keys in the list of the agent instances."
+                )
+            self.service.number_of_agents = len(keys)
```

Comment:
here it's the same thing right
```
mapping = {kp["address"]: kp for kp in keys}
keys = [mapping[address] for address in self.agent_instances]
```
which will error out in case of a missing key so no check needed either

Reply:
don't do that, do this: https://github.com/valory-xyz/open-autonomy/pull/1194#discussion_r946522585